File: system/governance/CSRE_validator.py
import logging
from typing import Dict, Any, Protocol, Type, Final
from pydantic import BaseModel, ValidationError
import importlib

# --- Core Exceptions ---

class GovernanceHalt(Exception):
    """Base exception for reconciliation failures, preventing CRoT activation."""
    pass

class SchemaIntegrityBreach(GovernanceHalt):
    """Raised when ACVD structure fails Pydantic schema validation."""
    pass

class PolicyLogicError(GovernanceHalt):
    """Raised when validated data violates critical logical constraints (e.g., negative metrics)."""
    pass

class ConfigurationLoadError(GovernanceHalt):
    """Raised if necessary Pydantic models cannot be imported or are malformed."""
    pass

# --- Constants & Configuration Keys ---
# Standard snake_case attribute name assumption for Pydantic models
ACVD_THRESHOLD_KEY: Final[str] = "acvd_threshold"

# NOTE: These constants define the dynamic location used by _load_model
POLICY_MODEL_PATH: Final[str] = ".governance_models"
POLICY_MODEL_NAME: Final[str] = "PolicyConfigurationModel"

# --- Policy Server Definition ---
class AbstractPolicyServer(Protocol):
    """Protocol defining the interface for fetching raw configuration data."""
    def fetch_acvd(self) -> Dict[str, Any]:
        """Must return the raw configuration state dictionary (ACVD)."""
        ...

# --- Setup ---
logger = logging.getLogger('CSRE')
logger.setLevel(logging.INFO)

# Type alias for cleaner references to the Pydantic model class
PolicyModelType = Type[BaseModel]

class ConfigStateReconciliationEngine:
    """Validates and reconciles configuration state (e.g., ACVD files) before handing off to CRoT.
    Enforces mandatory structured validation (Pydantic), leveraging V2 features.
    """

    def __init__(self, policy_server: AbstractPolicyServer):
        self.pcs: AbstractPolicyServer = policy_server
        
        # Policy model must be loaded during initialization (fail-fast principle)
        # Note: If a GovernanceModelResolver were used, this method would be refactored.
        self._PolicyModel: PolicyModelType = self._load_model()
        
        logger.debug(f"CSRE initialized using model: {self._PolicyModel.__name__}")

    def _load_model(self) -> PolicyModelType:
        """Tries to dynamically import the necessary Pydantic configuration model."""
        try:
            # Resolving relative import based on expected package structure (system.governance)
            module = importlib.import_module(POLICY_MODEL_PATH, package="system.governance") 
            
            PolicyConfigurationModel = getattr(module, POLICY_MODEL_NAME)
            
            if not issubclass(PolicyConfigurationModel, BaseModel):
                 raise TypeError(f"Object {POLICY_MODEL_NAME} is not a Pydantic BaseModel.")

            return PolicyConfigurationModel
        
        except (ImportError, AttributeError, TypeError, ValueError) as e:
            logger.error(f"FATAL CONFIG: Cannot load policy model '{POLICY_MODEL_PATH}.{POLICY_MODEL_NAME}'. Error: {type(e).__name__}: {e}")
            raise ConfigurationLoadError(f"Dependency load failed for policy model: {e}")

    def pre_vet_policies(self) -> BaseModel:
        """Fetches critical configs, verifies schema integrity, and checks fundamental policy logic.
           Returns the validated Pydantic model on success."""
        
        try:
            raw_data = self.pcs.fetch_acvd()

            # 1. Schema Validation 
            validated_model = self._validate_schema(raw_data)
            
            # 2. Advanced Logic Check 
            self._check_threshold_logic(validated_model)

        except GovernanceHalt:
            # Catch known system-halting errors
            raise
        except Exception as e:
             # Catch unexpected exceptions (e.g., connection errors, severe runtime issues)
            logger.critical(f"UNEXPECTED HALT: Internal processing error during pre-vetting: {type(e).__name__}: {e}", exc_info=True)
            raise GovernanceHalt(f"Unexpected pre-vet failure: {type(e).__name__}")

        logger.info(f"ACVD policies successfully validated against {self._PolicyModel.__name__}. Ready for CRoT.")
        return validated_model

    def _validate_schema(self, raw_data: Dict[str, Any]) -> BaseModel:
        """Enforces Pydantic schema validation using the pre-loaded Policy Model (V2 method)."""
        
        try:
            validated_model = self._PolicyModel.model_validate(raw_data) 
            return validated_model
        except ValidationError as e:
            # Extract location and message for clear debugging
            detailed_errors = [f"loc={'/'.join(map(str, err['loc']))}, msg={err['msg']}" for err in e.errors()]
            error_message = f"Structured Validation Failed ({self._PolicyModel.__name__}). Errors:\n{'; '.join(detailed_errors)}"
            logger.warning(error_message)
            raise SchemaIntegrityBreach(error_message)

    def _check_threshold_logic(self, validated_data: BaseModel) -> bool:
        """Ensures that critical governance values meet logical constraints (e.g., non-negativity)."""
        
        try:
            threshold = getattr(validated_data, ACVD_THRESHOLD_KEY)
        except AttributeError:
             raise PolicyLogicError(f"Validation failure: Model {self._PolicyModel.__name__} passed schema check but is missing critical logic attribute: {ACVD_THRESHOLD_KEY}.")

        if not isinstance(threshold, (int, float)):
            # Secondary check for type integrity against the fetched value
            raise PolicyLogicError(f"ACVD Threshold value ({threshold}) failed mandatory numeric typing check (was {type(threshold).__name__}).")

        if threshold < 0:
             raise PolicyLogicError(f"ACVD Threshold detected as negative ({threshold}). TEMM constraint violation.")
        
        return True

File: governance/CSRE_validator.js
// /governance/CSRE_validator.js

// Objective 2: UNIFIER Protocol - Export all functional logic

/**
 * GOVERNANCE: Standard Laws.
 * UNIFIER_REF: Target Kernel.
 *
 * This module enforces mandatory structured validation (Schema/Logic)
 * for configuration state before system activation (CRoT).
 */

// --- Core Exceptions (Mapped from Python) ---

class GovernanceHalt extends Error {
    constructor(message) {
        super(message);
        this.name = 'GovernanceHalt';
    }
}

class SchemaIntegrityBreach extends GovernanceHalt {
    constructor(message) {
        super(message);
        this.name = 'SchemaIntegrityBreach';
    }
}

class PolicyLogicError extends GovernanceHalt {
    constructor(message) {
        super(message);
        this.name = 'PolicyLogicError';
    }
}

class ConfigurationLoadError extends GovernanceHalt {
    constructor(message) {
        super(message);
        this.name = 'ConfigurationLoadError';
    }
}

// --- Constants ---
const ACVD_THRESHOLD_KEY = "acvd_threshold";

// --- Abstract Policy Server Definition (Protocol mapping) ---

/**
 * @typedef {Object} AbstractPolicyServer
 * @property {() => Promise<Object>} fetch_acvd Must return the raw configuration state dictionary (ACVD).
 */

class ConfigStateReconciliationEngine {
    /**
     * @param {AbstractPolicyServer} policyServer
     * @param {Object} policySchema - Assumed external schema validator interface (must have a .validate(data) method).
     */
    constructor(policyServer, policySchema) {
        if (!policyServer || typeof policyServer.fetch_acvd !== 'function') {
             throw new ConfigurationLoadError("Policy Server must implement fetch_acvd.");
        }
        if (!policySchema || typeof policySchema.validate !== 'function') {
             // Assuming the injected schema provides a validation method
             throw new ConfigurationLoadError("Policy Schema definition is invalid or missing validation method (.validate).");
        }

        this.pcs = policyServer;
        this.PolicyModel = policySchema; // Store the validation schema object

        console.debug("CSRE initialized.");
    }

    async preVetPolicies() {
        /**
         * Fetches critical configs, verifies schema integrity, and checks fundamental policy logic.
         * Returns the validated object on success.
         */
        
        try {
            const rawData = await this.pcs.fetch_acvd();

            // 1. Schema Validation 
            const validatedData = this._validateSchema(rawData);
            
            // 2. Advanced Logic Check 
            this._checkThresholdLogic(validatedData);
            
            console.info("ACVD policies successfully validated. Ready for CRoT.");
            return validatedData;

        } catch (e) {
            if (e instanceof GovernanceHalt) {
                throw e; // Catch known system-halting errors
            }
            // Catch unexpected exceptions
            console.error(`UNEXPECTED HALT: Internal processing error during pre-vetting: ${e.name}: ${e.message}`, e);
            throw new GovernanceHalt(`Unexpected pre-vet failure: ${e.name}`);
        }
    }

    /**
     * Enforces schema validation using the injected Policy Model.
     * @param {Object} rawData 
     * @returns {Object} Validated data
     */
    _validateSchema(rawData) {
        try {
            // Assume PolicyModel.validate returns { data: Object, error: null } or throws
            const validationResult = this.PolicyModel.validate(rawData);

            if (validationResult.error) {
                const errorDetail = JSON.stringify(validationResult.error);
                 throw new SchemaIntegrityBreach(`Structured Validation Failed. Details: ${errorDetail}`);
            }
            return validationResult.data; // Return the parsed/validated data
            
        } catch (e) {
            if (e instanceof SchemaIntegrityBreach) {
                throw e;
            }
            // Catch native validation library errors and wrap them
            throw new SchemaIntegrityBreach(`Validation failed due to internal error: ${e.message}`);
        }
    }

    /**
     * Ensures that critical governance values meet logical constraints.
     * @param {Object} validatedData 
     * @returns {boolean}
     */
    _checkThresholdLogic(validatedData) {
        if (!Object.prototype.hasOwnProperty.call(validatedData, ACVD_THRESHOLD_KEY)) {
             throw new PolicyLogicError(`Validation failure: Configuration model is missing critical logic attribute: ${ACVD_THRESHOLD_KEY}.`);
        }
        
        const threshold = validatedData[ACVD_THRESHOLD_KEY];

        if (typeof threshold !== 'number' || isNaN(threshold)) {
            // Secondary check for type integrity
            throw new PolicyLogicError(`ACVD Threshold value (${threshold}) failed mandatory numeric typing check (was ${typeof threshold}).`);
        }

        if (threshold < 0) {
             throw new PolicyLogicError(`ACVD Threshold detected as negative (${threshold}). TEMM constraint violation.`);
        }
        
        return true;
    }
}

module.exports = {
    ConfigStateReconciliationEngine,
    GovernanceHalt,
    SchemaIntegrityBreach,
    PolicyLogicError,
    ConfigurationLoadError,
    // Exporting exceptions for consumption by UNIFIER error handling
};
